File: packages/app/src/rate-limiter.ts
import type { RateLimiter, RateLimitResult } from "./types.js";

export class SimpleRateLimiter implements RateLimiter {
  private requests = new Map<string, number[]>();

  constructor(
    private requestsPerMinute: number = 60,
    private windowMs: number = 60000
  ) {}

  async checkLimit(userId: string): Promise<RateLimitResult> {
    const now = Date.now();
    const windowStart = now - this.windowMs;

    // Get user's requests within window
    let userRequests = this.requests.get(userId) ?? [];
    userRequests = userRequests.filter((time) => time > windowStart);

    const count = userRequests.length;
    const allowed = count < this.requestsPerMinute;

    if (allowed) {
      userRequests.push(now);
      this.requests.set(userId, userRequests);
    }

    return {
      allowed,
      remaining: Math.max(0, this.requestsPerMinute - count - 1),
      resetAt: new Date(windowStart + this.windowMs),
    };
  }
}

File: crates/spectra-rs/src/rate_limiter.rs
use std::collections::HashMap;
use std::sync::Mutex;
use std::time::{Duration, Instant};

#[derive(Debug, Clone)]
pub struct RateLimitResult {
    pub allowed: bool,
    pub remaining: u32,
    pub window_end: Instant,
}

pub trait RateLimiter: Send + Sync {
    fn check_limit(&self, key: &str) -> RateLimitResult;
}

pub struct LocalRateLimiter {
    requests: Mutex<HashMap<String, Vec<Instant>>>,
    max_requests: u32,
    window: Duration,
}

impl LocalRateLimiter {
    pub fn new(max_requests: u32, window: Duration) -> Self {
        Self {
            requests: Mutex::new(HashMap::new()),
            max_requests,
            window,
        }
    }

    pub fn with_defaults() -> Self {
        Self::new(60, Duration::from_secs(60))
    }
}

impl RateLimiter for LocalRateLimiter {
    fn check_limit(&self, key: &str) -> RateLimitResult {
        let now = Instant::now();
        let window_start = now - self.window;

        let mut map = self.requests.lock().unwrap();
        let timestamps = map.entry(key.to_string()).or_default();

        timestamps.retain(|t| *t > window_start);

        let count = timestamps.len() as u32;
        let allowed = count < self.max_requests;

        if allowed {
            timestamps.push(now);
        }

        RateLimitResult {
            allowed,
            remaining: self.max_requests.saturating_sub(count + 1),
            window_end: now + self.window,
        }
    }
}
